compiler/fir/resolve/src/org/jetbrains/kotlin/fir/types/ConeTypeCompatibilityChecker.kt

```diff
@@ -0,0 +1,437 @@
+    }
+
+    /**
+     * @param compatibilityUpperBound the max compatibility result that can be returned by this method. For example, if this is set to
+     * [Compatibility.SOFT_INCOMPATIBLE], then even if the given bounds don't match the hard way (for example, incompatible primitives) the
+     * method should still return [Compatibility.SOFT_INCOMPATIBLE]. This is useful for checking type parameters since we don't want to
+     * dictate what semantics a type parameter may have in user code. In other words, if user wants to compare `MyCustom<out String>` with
+     * `MyCustom<out Int>`, we let them do so since we do not know what class `MyCustom` uses the type parameter for. Empty containers are
+     * another example: `emptyList<Int>() == emptyList<String>()`.
+     */
+    private fun ConeInferenceContext.areCompatible(
+        upperBounds: Set<ConeClassLikeType>,
+        lowerBounds: Set<ConeClassLikeType>,
+        compatibilityUpperBound: Compatibility
+    ): Compatibility {
+        val upperBoundClasses: Set<ConcreteClass> = upperBounds.mapNotNull { it.toConcreteClass(this) }.toSet()
+
+        // Following if condition is an optimization: if we ignore the subtyping relation and treat all upper bounds as unrelated
+        // classes/interfaces, yet the types are deemed compatible for sure, then we just bail out early.
+        if (lowerBounds.isEmpty() &&
+            (upperBounds.size < 2 ||
+                    areClassesOrInterfacesCompatible(upperBoundClasses, compatibilityUpperBound) == Compatibility.COMPATIBLE)
+        ) {
+            return Compatibility.COMPATIBLE
+        }
+
+        val leafClassesOrInterfaces = computeLeafClassesOrInterfaces(upperBoundClasses)
+        areClassesOrInterfacesCompatible(leafClassesOrInterfaces, compatibilityUpperBound)?.let { return it }
+
+        // Check if the range formed by upper bounds and lower bounds is empty.
+        if (!lowerBounds.all { lowerBoundType ->
+                val classesSatisfyingLowerBounds =
+                    lowerBoundType.toConcreteClass(this)?.thisAndAllSuperConcreteClasses ?: emptySet()
+                leafClassesOrInterfaces.all { it in classesSatisfyingLowerBounds }
+            }
+        ) {
+            return compatibilityUpperBound
+        }
+
+        if (upperBounds.size < 2) return Compatibility.COMPATIBLE
+
+        // Base types are compatible. Now we check type parameters.
+
+        val typeArgumentMapping = mutableMapOf<FirTypeParameterRef, BoundTypeArguments>().apply {
+            for (type in upperBounds) {
+                collectTypeArgumentMapping(type, this@areCompatible, compatibilityUpperBound)
+            }
+        }
+        var result = Compatibility.COMPATIBLE
+        for (compatibility in typeArgumentMapping.values.asSequence()
+            .map { (upper, lower, compatibilityUpperBound) -> areCompatible(upper, lower, compatibilityUpperBound) }) {
```

Comment:
It'd be better to avoid this wrapped line in the middle of `for`.

Actually... you can simplify these with `maxByOrNull` and `Compatibility.ordinal`, something like:
```kt
val result = typeArgumentMapping.values...
    .map { ... }
    .maxByOrNull { it.ordinal }
    ?: compatibilityUpperBound
return if (result < compatibilityUpperBound) result else compatibilityUpperBound
```

Reply:
I can give the sequence a name. But using the for loop was intended so that the execution can stop as soon as some incompatibility is detected.